core/Segment.php

```diff
@@ -180,17 +182,45 @@ public function getSegmentExpression()
         return $this->segmentExpression;
     }
 
+    /**
+     * @throws Exception
+     */
     private function getAvailableSegments()
     {
+        // start lazy cache
+        $cache = PiwikCache::getLazyCache();
+
+        //covert lazy cache to lock ID
+        $lockId = implode(",", $this->idSites);
+
+        $cacheId = SettingsPiwik::getPiwikInstanceId().$lockId;
+
+       //fetch cache lockId
+        $this->availableSegments = $cache->fetch($lockId);
+
         // segment metadata
+        // restart cache if load is empty
         if (empty($this->availableSegments)) {
-            $this->availableSegments = Request::processRequest('API.getSegmentsMetadata', array(
-                'idSites' => $this->idSites,
-                '_hideImplementationData' => 0,
-                'filter_limit' => -1,
-                'filter_offset' => 0,
-                '_showAllSegments' => 1,
-            ), []);
+            //init lock
+            $lock = new LoaderLock($lockId);
```

Comment:
@peterhashair is there a particular reason the lock is used? It be fine if multiple processes try to populate this cache at the same time.